src/runtime-rs/crates/hypervisor/src/qemu/qmp.rs

```diff
@@ -235,6 +235,62 @@ impl Qmp {
 
         Ok(())
     }
+
+    pub fn hotunplug_memory(&mut self, size: i64) -> Result<()> {
+        let frontend = self
+            .qmp
+            .execute(&qapi_qmp::query_memory_devices {})?
+            .into_iter()
+            .find(|memdev| {
+                if let qapi_qmp::MemoryDeviceInfo::dimm(dimm_info) = memdev {
+                    let dimm_id = match dimm_info.data.id {
+                        Some(ref id) => id,
+                        None => return false,
+                    };
+                    if dimm_info.data.hotpluggable
+                        && dimm_info.data.hotplugged
+                        && dimm_info.data.size == size
+                        && dimm_id.starts_with("frontend-to-hotplugged-")
+                    {
+                        return true;
+                    }
+                }
+                false
+            });
+
+        if let Some(frontend) = frontend {
+            if let qapi_qmp::MemoryDeviceInfo::dimm(frontend) = frontend {
+                info!(sl!(), "found frontend to hotunplug: {:#?}", frontend);
+
+                let frontend_id = match frontend.data.id {
+                    Some(id) => id,
+                    // This shouldn't happen as it was checked by find() above already.
+                    None => return Err(anyhow!("memory frontend to hotunplug has empty id")),
+                };
+
+                let backend_id = match frontend_id.strip_prefix("frontend-to-") {
+                    Some(id) => id.to_owned(),
+                    // This shouldn't happen as it was checked by find() above already.
+                    None => {
+                        return Err(anyhow!(
+                        "memory backend to hotunplug has id that doesn't have the expected prefix"
+                    ))
+                    }
+                };
+
+                self.qmp.execute(&qmp::device_del { id: frontend_id })?;
```

Comment:
I have the impression that this can punch a hole in the index range and cause next hotplug to try to reuse an existing id... but I could be wrong. In any case, not to be fixed now since this is experimental.

Reply:
I _think_ you're right but I don't think it would be a problem.  If a hole is created by removing a memdev then the id is not "existing" anymore.  The whole idea here is to have device names that 1) allow us to tell devices that we created (by hotplugging) from any other dimms, and 2) allow to create several memdevs with names that are still unique.  The id serves the second point.

And yeah, that all might be moot since hotunplugging seems to fail fairly consistently currently and if there isn't a successful effort in the future to support hotunplugging in the agent this function might even be removed altogether.

Reply:
Yeah as discussed some times ago, memory hot-unplug requires some care from the guest kernel : it shouldn't place some long living and not movable structure in hot-plugged memory.